src/tir/transforms/lower_sparse_tir.cc

```diff
@@ -0,0 +1,306 @@
+          }
+        } else {
+          LOG(FATAL) << "Cannot reach here";
+        }
+      }
+
+      // Stage 2. Accumulate the lowered index.
+      lowered_index =
+          AccumulateLowerIndex(std::move(lowered_index), sp_buffer, i, std::move(sp_index));
+    }
+
+    return lowered_index;
+  }
+
+  PrimExpr AccumulateLowerIndex(PrimExpr prev_lowered_index, const SparseBuffer& sp_buffer, int dim,
+                                PrimExpr index) {
+    const Axis& axis = sp_buffer->axes[dim];
+    if (axis->IsInstance<DenseFixedAxisNode>() || axis->IsInstance<SparseFixedAxisNode>()) {
+      return ana_.Simplify(std::move(prev_lowered_index) * axis->length + std::move(index));
+    } else if (const auto* dv_axis = axis.as<DenseVariableAxisNode>()) {
+      return ana_.Simplify(
+          Add(BufferLoad(dv_axis->indptr, {std::move(prev_lowered_index)}), std::move(index)));
+    } else if (const auto* sv_axis = axis.as<SparseVariableAxisNode>()) {
+      return ana_.Simplify(
+          Add(BufferLoad(sv_axis->indptr, {std::move(prev_lowered_index)}), std::move(index)));
+    }
+    LOG(FATAL) << "Cannot reach here";
+    throw;
+  }
+
+  PrimExpr GetDenseValue(const SpIterVarNode* sp_iter) {
+    SpIterKind kind = sp_iter->kind;
+    CHECK(kind == SpIterKind::kSparseFixed || kind == SpIterKind::kSparseVariable);
+    Axis iterated_axis = sp_iter->axis.value();
+
+    std::pair<SparseBuffer, int> depended_pair = dependency_map_[GetRef<SpIterVar>(sp_iter)];
+    Array<SpIterVar> buffer_access_iters = buffer_access_map_[depended_pair.first];
+    int n_depended = depended_pair.second;
+
+    Array<PrimExpr> depended_iters{buffer_access_iters.begin(),
+                                   buffer_access_iters.begin() + n_depended};
+    PrimExpr lowered_indices = LowerIndices(depended_pair.first, depended_iters);
+
+    if (kind == SpIterKind::kSparseFixed) {
+      return BufferLoad(Downcast<SparseFixedAxis>(iterated_axis)->indices,
+                        {std::move(lowered_indices)});
+    } else {
+      return BufferLoad(Downcast<SparseVariableAxis>(iterated_axis)->indices,
+                        {std::move(lowered_indices)});
+    }
+  }
+
+  PrimExpr VisitExpr_(const SparseBufferLoadNode* load) final {
+    PrimExpr lowered_indices = LowerIndices(load->buffer, load->indices);
+    return BufferLoad(load->buffer->data, {std::move(lowered_indices)});
+  }
+
+  Stmt VisitStmt_(const SparseBufferStoreNode* store) final {
+    PrimExpr value = ExprMutator::VisitExpr(store->value);
+    PrimExpr lowered_indices = LowerIndices(store->buffer, store->indices);
+    return BufferStore(store->buffer->data, std::move(value), {std::move(lowered_indices)});
+  }
+
+  BufferAccessMap buffer_access_map_;
+  DependencyMap dependency_map_;
+  arith::Analyzer ana_;
+};
+
+PrimFunc LowerSparseTIR(PrimFunc f) {
+  // Only apply this pass to TIR that is not from TE schedules
+  if (!IsFromLegacyTESchedule(f)) {

```

Comment:
Most of the time when we write TIR passes, we don't take TE cases into consideration. For example, we may assume that the `body` field of each `PrimFunc` is a `BlockRealize`. However, this property doesn't hold for `PrimFunc`s lowered from TE schedules. Therefore such a check was introduced several months ago, and the purpose of this check is to prevent `PrimFunc`s lowered from TE schedules run into the check to avoid unexpected behavior.

> I suppose we can use a global attribute to determine whether to trigger this pass or not.

I'm not sure whether the attribute is necessary. IMO this pass doesn't change any `PrimFunc` that has no `SparseBufferLoad/Store`. Therefore it's safe for all TIR `PrimFunc`s to run into this pass 👀.

Reply:
Make sense to me.